File: 2026/dp/min_ascii_delete_712.js
/**
 * @param {string} s1
 * @param {string} s2
 * @return {number}
 */
var minimumDeleteSum = function (s1, s2) {
    const m = s1.length;
    const n = s2.length;
    const t = Array.from({ length: m + 1 }, () => Array(n + 1).fill(-1));
    const solve = (i, j) => {
        if (i >= m && j >= n) {
            return 0;
        }
        if (t[i][j] !== -1) {
            return t[i][j];
        }
        if (i >= m) {
            return t[i][j] = s2.charCodeAt(j) + solve(i, j + 1);
        } else if (j >= n) {
            return t[i][j] = s1.charCodeAt(i) + solve(i + 1, j);
        }
        let delete_i = s1.charCodeAt(i) + solve(i + 1, j);
        let delete_j = s2.charCodeAt(j) + solve(i, j + 1);
        let options = [delete_i, delete_j];
        if (s1[i] === s2[j]) {
            options.push(solve(i + 1, j + 1));
        }
        t[i][j] = Math.min(...options);
        return t[i][j];
    };
    return solve(0, 0);
};

File: 2026/dp/min_ascii_delete_712.py
class Solution:
    def __init__(self):
        self.m = 0
        self.n = 0
        self.t = []

    def solve(self, s1, s2, i, j):
        if i >= self.m and j >= self.n:
            return 0

        if self.t[i][j] != -1:
            return self.t[i][j]

        if i >= self.m:
            self.t[i][j] = ord(s2[j]) + self.solve(s1, s2, i, j + 1)
            return self.t[i][j]
        elif j >= self.n:
            self.t[i][j] = ord(s1[i]) + self.solve(s1, s2, i + 1, j)
            return self.t[i][j]

        if s1[i] == s2[j]:
            self.t[i][j] = self.solve(s1, s2, i + 1, j + 1)
            return self.t[i][j]

        take_s1_i = ord(s1[i]) + self.solve(s1, s2, i + 1, j)
        take_s2_j = ord(s2[j]) + self.solve(s1, s2, i, j + 1)

        self.t[i][j] = min(take_s1_i, take_s2_j)
        return self.t[i][j]

    def minimumDeleteSum(self, s1: str, s2: str) -> int:
        self.m = len(s1)
        self.n = len(s2)
        self.t = [[-1] * (self.n + 1) for _ in range(self.m + 1)]
        return self.solve(s1, s2, 0, 0)
